core/src/main/scala/org/apache/spark/sql/SnappyContext.scala

```diff
@@ -970,39 +927,70 @@ object SnappyContext extends Logging {
       }
     }
   }
+  private[sql] def initGlobalSnappyContext(sc: SparkContext) = {
+    if (!_globalSNContextInitialized) {
+      contextLock.synchronized {
+        if (!_globalSNContextInitialized) {
+          invokeServices(sc)
+          sc.addSparkListener(new SparkContextListener)
+          sc.ui.foreach(new SnappyStatsTab(_))
+          _globalSNContextInitialized = true
+        }
+      }
+    }
+  }
+
+  private class SparkContextListener extends SparkListener {
+    override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd): Unit = {
+      stop()
+    }
+  }
+
+  private def invokeServices(sc: SparkContext): Unit = {
+    SnappyContext.getClusterMode(sc) match {
+      case SnappyEmbeddedMode(_, _) =>
+        // NOTE: if Property.jobServer.enabled is true
+        // this will trigger SnappyContext.apply() method
+        // prior to `new SnappyContext(sc)` after this
+        // method ends.
+        ToolsCallbackInit.toolsCallback.invokeLeadStartAddonService(sc)
+        SnappyDaemons.start(sc)
+      case SnappyShellMode(_, _) =>
+        ServiceUtils.invokeStartFabricServer(sc, hostData = false)
+        SnappyDaemons.start(sc)
+      case ExternalEmbeddedMode(_, url) =>
+        SnappyContext.urlToConf(url, sc)
+        ServiceUtils.invokeStartFabricServer(sc, hostData = false)
+        SnappyDaemons.start(sc)
+      case LocalMode(_, url) =>
+        SnappyContext.urlToConf(url, sc)
+        ServiceUtils.invokeStartFabricServer(sc, hostData = true)
+        SnappyDaemons.start(sc)
+      case _ => // ignore
+    }
+  }
 
   /**
-   * Shut down and cleanup the SparkContext and SnappyData artifacts.
-   * Prefer this over SparkContext.stop() when dealing with SnappyContext.
-   * <p>
-   * This method is not synchronized and is required to be executed by a single
-   * thread in a "quiet" state when no other threads are performing operations.
-   */
-  def stop(): Unit = {
+    * Shut down and cleanup and SnappyData artifacts.
+    */
+  private def stop(): Unit = {
```

Comment:
See my comment before about adding "stopSparkContext" flag to this. Will make it similar to StreamingContext and can be invoked with argument as true from tests etc.